File: soln-sliding-window-problems/subarrays-with-k-different-integers.java
/*
https://leetcode.com/problems/subarrays-with-k-different-integers
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public int subarraysWithKDistinct(int[] A, int K) {
        return atMostK(A,K)-atMostK(A,K-1);
    }
        
    public int atMostK(int[] A, int K){ 
        HashMap<Integer,Integer> map=new HashMap<>();
        
        int i=0;
        int j=0;
        int count=0;
        
        while(j<A.length){
            map.put(A[j],map.getOrDefault(A[j],0)+1);
            
            while(map.size()>K){
                map.put(A[i],map.get(A[i])-1);
                if(map.get(A[i])==0){
                    map.remove(A[i]);
                }
                i++;
            }
            count+=j-i+1;
            j++;
        }
        return count;
    }
}

File: Python/soln-sliding-window-problems/subarrays-with-k-different-integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
     
        def window(nums, k):
            i,j=0,0
            count=0

            count_dict={}

            while j<len(nums):
                if nums[j] not in count_dict:
                    count_dict[nums[j]]=0
                count_dict[nums[j]]+=1

                while(len(count_dict))>k:
                    count_dict[nums[i]]-=1

                    if count_dict[nums[i]]==0:
                        count_dict.pop(nums[i],None)
                    i+=1

                count+=j-i+1
                j+=1

            return count
        
        return window(nums,k)-window(nums,k-1)
            